skiko/src/awtMain/kotlin/org/jetbrains/skiko/swing/SwingOffscreenDrawer.kt

```diff
@@ -0,0 +1,274 @@
+            // loss if the image is drawn via Graphics.drawImage(image, <srcRect>, <dstRect>)
+            // due to scaling in Graphics. To avoid that, the image should be drawn directly via
+            // Graphics.drawImage(image, 0, 0) on the unscaled Graphics.
+            val gScaleX = tx.scaleX
+            val gScaleY = tx.scaleY
+            tx.scale(1 / gScaleX, 1 / gScaleY)
+            tx.translate(dx1 * gScaleX, dy1 * gScaleY)
+            dy1 = 0
+            dx1 = 0
+            invG = g1.create() as Graphics2D
+            g1 = invG
+            invG.transform = tx
+        }
+
+        try {
+            var dw = dw
+            var dh = dh
+            if (invG != null && hasDestinationSize) {
+                dw = scaleSize(dw, scale1)
+                dh = scaleSize(dh, scale1)
+            }
+            doDraw(
+                op = op,
+                image = image,
+                invG = invG,
+                hasDestinationSize = hasDestinationSize,
+                dw = dw,
+                dh = dh,
+                sourceBounds = srcBounds,
+                userWidth = userWidth,
+                userHeight = userHeight,
+                g = g1,
+                dx = dx1,
+                dy = dy1,
+                observer = observer,
+                scale = scale1
+            )
+        } finally {
+            invG?.dispose()
+        }
+    }
+
+    private fun scaleSize(size: Int, scale: Double) = (size * scale).roundToInt()
+
+    @Suppress("NAME_SHADOWING")
+    private fun doDraw(
+        op: BufferedImageOp?,
+        image: Image,
+        invG: Graphics2D?,
+        hasDestinationSize: Boolean,
+        dw: Int,
+        dh: Int,
+        sourceBounds: Rectangle?,
+        userWidth: Int,
+        userHeight: Int,
+        g: Graphics,
+        dx: Int,
+        dy: Int,
+        observer: ImageObserver?,
+        scale: Double
+    ) {
+        var image = image
+        if (op != null && image is BufferedImage) {
+            image = op.filter(image, null)
+        }
+
+        when {
+            sourceBounds != null -> {
+                fun size(size: Int) = scaleSize(size, scale)
+
+                val sx = size(sourceBounds.x)
+                val sy = size(sourceBounds.y)
+                val sw = if (sourceBounds.width >= 0) size(sourceBounds.width) else size(userWidth) - sx
+                val sh = if (sourceBounds.height >= 0) size(sourceBounds.height) else size(userHeight) - sy
+
+                var dw = dw
+                var dh = dh
+                if (!hasDestinationSize) {
+                    dw = size(userWidth)
+                    dh = size(userHeight)
+                }
+                g.drawImage(/* img = */ image,
+                    /* dx1 = */ dx, /* dy1 = */ dy, /* dx2 = */ dx + dw, /* dy2 = */ dy + dh,
+                    /* sx1 = */ sx, /* sy1 = */ sy, /* sx2 = */ sx + sw, /* sy2 = */ sy + sh,
+                    /* observer = */ observer
```

Comment:
Why are labels commented out?

Reply:
This is a way how we can use naming parameters when calling java function